tips/tip-0008/tip-0008.md

```diff
@@ -51,9 +51,22 @@ The state of the VM is fully determined by a bunch of things, divisible into the
 
 The VM is initialized with the program hash on the bottom of its opstack. By making these five elements **read-only**, we get program integrity from opstack integrity and collision-resistance.
 
-The only way in which the program hash could be overwritten is through `swap` instructions when the opstack size is 21 or less. 
+The only way in which the program hash could be overwritten is through `swap`, `pick`, and `place` instructions when the opstack size is 21 or less. 
 
-**Constraints.** Todo.
+**Constraints.** To prevent the program hash from being overwritten by stack manipulation instructions when the opstack size is 21 or less, we need to ensure that these operations cannot target the program hash elements (st11 through st15) when the stack is shallow enough to expose them.
+
+Let $n$ be the current opstack size and let $i$ be the target index. The constraint is:
+
+$$\text{stack\_manip\_indicator} \cdot (21 - n) \cdot \text{indicator}_{i \geq 11} = 0$$
```

Comment:
I'm afraid this constraint is not complete. Counterexample: the instruction is `swap 15` when $n = 22$. The constraint evaluates to $P-1 \neq 0$ meaning that this particular instruction in this particular context is not allowed. But it *should be* allowed because when $n = 22$ there is no risk of disturbing the program hash.